crates/stages/src/stages/headers.rs

```diff
@@ -74,63 +74,57 @@ impl<DB: Database, D: HeaderDownloader, C: Consensus, H: HeadersClient, S: Statu
-        let mut stream =
-            self.downloader.stream(head.clone(), tip).chunks(self.commit_threshold as usize);
+        let downloaded_headers: Result<Vec<SealedHeader>, DownloadError> =
+            self.downloader.stream(head.clone(), tip).try_collect().await;
         // The stage relies on the downloader to return the headers
         // in descending order starting from the tip down to
         // the local head (latest block in db)
-        while let Some(headers) = stream.next().await {
-            match headers.into_iter().collect::<Result<Vec<_>, _>>() {
-                Ok(res) => {
-                    info!(target: "sync::stages::headers", len = res.len(), "Received headers");
-                    self.metrics.headers_counter.increment(res.len() as u64);
-
-                    // Perform basic response validation
-                    self.validate_header_response(&res)?;
-                    let write_progress =
-                        self.write_headers::<DB>(tx, res).await?.unwrap_or_default();
-                    current_progress = current_progress.max(write_progress);
-                }
-                Err(e) => {
-                    self.metrics.update_headers_error_metrics(&e);
-                    match e {
-                        DownloadError::Timeout => {
-                            warn!(target: "sync::stages::headers", "No response for header request");
-                            return Err(StageError::Recoverable(DownloadError::Timeout.into()))
-                        }
-                        DownloadError::HeaderValidation { hash, error } => {
-                            error!(target: "sync::stages::headers", ?error, ?hash, "Validation error");
-                            return Err(StageError::Validation { block: stage_progress, error })
-                        }
-                        error => {
-                            error!(target: "sync::stages::headers", ?error, "Unexpected error");
-                            return Err(StageError::Recoverable(error.into()))
-                        }
+        match downloaded_headers {
+            Ok(res) => {
+                info!(target: "sync::stages::headers", len = res.len(), "Received headers");
+                self.metrics.headers_counter.increment(res.len() as u64);
+                let max_received_block_number = res[0].number;
+
+                // Perform basic response validation
+                self.validate_header_response(&res)?;
+
+                let write_progress = self
+                    .write_headers::<DB>(tx, res, self.commit_threshold)
+                    .await?
+                    .unwrap_or_default();
+                current_progress = current_progress.max(write_progress);
+                // Write total difficulty values after headers have been inserted
+                debug!(target: "sync::stages::headers", head = ?head.hash(), "Writing total difficulty");
+                self.write_td::<DB>(tx, &head)?;
```

Comment:
this is not correct: total difficulty has to be written onwards since it's a cumulative sum based on the previous value